File: ts-sdk/src/services/pipeline.ts
import { LogRecord } from '../core/record';
import { LogEnricher, LogSampler, LogHandler } from '../core/interfaces';

export class PipelineService {
  constructor(
    private enrichers: LogEnricher[],
    private sampler: LogSampler | null,
    private handlers: LogHandler[]
  ) {}

  async run(record: LogRecord): Promise<void> {
    // 1. Enrich
    let current = record;
    for (const enricher of this.enrichers) {
      current = enricher.enrich(current);
    }

    // 2. Sample
    if (this.sampler && !this.sampler.shouldSample(current)) {
      return;
    }

    // 3. Handle
    for (const handler of this.handlers) {
      handler.handle(current);
    }
  }
}

File: py-sdk/mylogger/services/pipeline.py
from typing import List, Optional
from ..core.record import LogRecord
from ..core.interfaces import LogEnricher, LogHandler, LogSampler

class PipelineService:
    def __init__(self, 
                 enrichers: List[LogEnricher],
                 sampler: Optional[LogSampler],
                 handlers: List[LogHandler]):
        self.enrichers = enrichers
        self.sampler = sampler
        self.handlers = handlers

    def run(self, record: LogRecord) -> None:
        # 1. Enrich
        for enricher in self.enrichers:
            record = enricher.enrich(record)

        # 2. Sample
        if self.sampler and not self.sampler.should_sample(record):
            return

        # 3. Handle
        for handler in self.handlers:
            handler.handle(record)
